### PR TITLE
:sparkles: Real time censoring with file output and playback

### LINKED ISSUE #134: Real-Time Chunk Censoring
- [x] Create wiki page for real-time installation
  - Break it up based on OS
- [x] Put wiki link in the README
- [ ] Grab the system's audio into chunks and pass it through the censor process
- [ ] Console print animation
- [ ] Maximize the system's output volume

### LINKED ISSUE #135: Real-Time File Output
- [ ] Create a (large) file based on the real-time audio
  - Create a new file after 10 minutes
- [ ] Stop adding to the file if there's is more than 3 seconds of silence
  - Once there's sound again, starting writing to a new file
- [ ] Print when a new file is created

### PR DESCRIPTION
**Pull request by levin-noro**
Closes #134
Closes #135 

Initial version for review and testing.
Will cleanup print statements and do any refactoring in subsequent versions.

### CODE REVIEW

File: cleansio/censor/censor_realtime_mac.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        with self.playback_lock:
+            if len(self.playback_queue) > 0:
+                outdata[:] = self.playback_queue.pop(0)
+            else:
+                outdata.fill(0)
+
+    def finished_callback(self):
+        self.create_clean_file(self.audio_file)
+
+    def run(self):
```
Comment (PatrickDuncan) [inline cleansio/censor/censor_realtime_mac.py]: You should break up this huge function in a few methods

File: cleansio/censor/censor_realtime_mac.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+                with open(overlapping_accuracy_path, 'wb') as chunk_file:
+                    overlapping_accuracy_chunk = improve_accuracy(overlapping_chunk)
+                    convert_and_write_chunk(overlapping_accuracy_chunk,chunk_file,'wav')
+
+                clean_chunk = self.censor_audio_chunk(file_path)
+                clean_chunk_filepath = create_temp_dir() + 'clean_chunk.wav'
+                clean_chunk.export(clean_chunk_filepath, format='wav')
+
+                # Read and convert it to frames
+                clean_frames, sample_rate = sf.read(clean_chunk_filepath,dtype='float32',fill_value=0.0,frames=int(44100*5),always_2d=True)
```
Comment (PatrickDuncan) [inline cleansio/censor/censor_realtime_mac.py]: Spaces between params

File: cleansio/censor/censor_realtime_mac.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+            print('\nInterrupted by user')
+            CensorRealtimeMac.running = False
+        except Exception as e:
+            print(type(e).__name__ + ': ' + str(e))
+            CensorRealtimeMac.running = False
+
+        # processing_thread.join()
+
+    def callback(self, indata, outdata, frames, time, status):
+        if status:
```
Comment (PatrickDuncan) [inline cleansio/censor/censor_realtime_mac.py]: Remove debug prints?

File: cleansio/censor/censor_realtime_mac.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+            print(status)
+
+        #add to processing_queue
+        with self.processing_lock:
```
Comment (PatrickDuncan) [inline cleansio/censor/censor_realtime_mac.py]: `# Comments should looks like this`

File: cleansio/censor/censor_realtime_mac.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+        processing_thread = threading.Thread(target=self.run)
+        processing_thread.daemon = True
+        processing_thread.start()
+
+        try:
+            # listen from Soundflower, play to speakers
+            with sd.Stream(device=(2, 1),
+                       samplerate=samplerate, blocksize=int(samplerate*duration),
+                       channels=1, callback=self.callback, finished_callback=self.finished_callback):
+                print('#' * 80)
```
Comment (PatrickDuncan) [inline cleansio/censor/censor_realtime_mac.py]: Remove debugging prints

File: cleansio/censor/censor_realtime_mac.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
+""" Censors audio chunks in a continuous stream """
+
+from .censor import Censor
+from utils import create_env_var, create_temp_dir, append_before_ext, \
+    time_filename, MacUtil
+from audio import improve_accuracy, convert_audio_segment, \
+    convert_and_write_chunk, read_and_convert_audio
+from pathlib import Path
+from pydub import AudioSegment
```
Comment (PatrickDuncan) [inline cleansio/censor/censor_realtime_mac.py]: Library imports, then our imports. Linting will tell you anyways

File: cleansio/censor/censor.py [hunk 7]
```diff
@@ -32,24 +43,40 @@ def __mute_explicits(self, file_path, audio_segment, timestamps):
                 audio_segment = self.__mute_explicit(audio_segment, stamp)
                 muted = True
                 chunk_index = int(file_path.split('-')[-1].split('.')[0])
+                print('Found ' + str(stamp) + ' at index ' + str(chunk_index))
```
Comment (PatrickDuncan) [inline cleansio/censor/censor.py]: Remove debugging

**Review by PatrickDuncan — CHANGES_REQUESTED** [review]

File: cleansio/censor/censor_realtime_mac.py (hunk 5 above)
Comment (levin-noro) [reply cleansio/censor/censor_realtime_mac.py]: This print provides useful instructions

File: cleansio/censor/censor_realtime_mac.py [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+        return recorded_chunk, file_path
+
+    def __convert_clean_chunk_to_frames(self, chunk):
+        clean_chunk_filepath = self.directory + 'clean_chunk.wav'
+        chunk.export(clean_chunk_filepath, format='wav')
+        clean_frames, _ = sf.read(clean_chunk_filepath, dtype='float32', fill_value=0.0, frames=int(self.samplerate*self.duration), always_2d=True)
+        return clean_frames
+
+    def __create_overlapping_chunk(self, chunk1, chunk2, file_path):
+        overlapping_chunk = chunk1[2500:] + chunk2[:2500]
```
Comment (levin-noro) [inline cleansio/censor/censor_realtime_mac.py]: For whatever reason, this results in an overlapping_chunk that is chunk1[2500:] + **chunk1**[:2500]. It makes absolutely no sense, and I can't figure out why its happening

File: cleansio/audio/convert.py [hunk 9]
```diff
@@ -21,9 +20,32 @@ def __create_converted_file(file_path, encoding):
...
+    """ Create a GCS AudioSegment """
+    audio_segment            \
+        .set_channels(1)     \
+        .set_sample_width(2) \
+        .set_frame_rate(__sample_rate(audio_segment))
+    return audio_segment
+
+def convert_and_write_chunk(chunk, file_path, encoding):
+    """ Create a GCS AudioSegment and write to the file path """
+    chunk.set_channels(1)      \
```
Comment (PatrickDuncan) [inline cleansio/audio/convert.py]: Inconsistent styling

File: cleansio/censor/censor.py [hunk 10]
```diff
@@ -1,40 +1,59 @@
...
         super().__init__()
         self.explicits = explicits
+        self.encoding = self.__encoding(output_encoding)
+        self.location = self.__location(output_location)
 
     def censor_audio_chunk(self, file_path):
         """ Common process to censor an audio chunk """
         audio_segment = AudioSegment.from_file(file_path)
         lyrics = self.__get_lyrics(file_path, audio_segment)
+        # print('lyrics = '+str(lyrics))
```
Comment (PatrickDuncan) [inline cleansio/censor/censor.py]: remove debug prints

File: cleansio/censor/censor_realtime_mac.py [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+        processing_thread.daemon = True
+        processing_thread.start()
+
+        try:
+            # listen from Soundflower, play to speakers
+            with sd.Stream(device=(2, 1),
+                       samplerate=self.samplerate, blocksize=int(self.samplerate*self.duration),
+                       channels=1, callback=self.callback, finished_callback=self.finished_callback):
+                print('#' * 80)
+                print('press Return to stop censoring')
```
Comment (PatrickDuncan) [inline cleansio/censor/censor_realtime_mac.py]: Capital p?

File: cleansio/utils/__init__.py [hunk 12]
```diff
@@ -4,6 +4,7 @@
 from .cli import setup_cli_args
 from .env import create_env_var
 from .files import create_temp_dir, file_name_no_ext, current_dir, \
-    relative_path, append_before_ext
+    relative_path, append_before_ext, time_filename
 from .numbers import gcs_time_to_ms, is_number, leading_zero
 from .constants import CHUNK_LEN
+from .mac import MacUtil
```
Comment (PatrickDuncan) [inline cleansio/utils/__init__.py]: needs newline

File: cleansio/utils/cleanup.py [hunk 13]
```diff
@@ -2,16 +2,21 @@
...
+from .mac import MacUtil
 
 # Cleans up files on normal or abnormal exit
 # The arguments are unused - they are only here to satisfy atexit.
 def cleanup(_sig_num=None, _cur_stack_frame=None):
     """ Removes temporary files """
-    remove_conversions()
-    remove_chunks()
+    # remove_conversions()
+    # remove_chunks()
```
Comment (PatrickDuncan) [inline cleansio/utils/cleanup.py]: ?

**Review by PatrickDuncan — CHANGES_REQUESTED** [review]

File: cleansio/censor/censor_realtime_mac.py (hunk 8 above)
Comment (levin-noro) [reply cleansio/censor/censor_realtime_mac.py]: Resolved by using indata.copy()

**Review by PatrickDuncan — DISMISSED** [review]

File: cleansio/censor/censor_realtime_mac.py [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+        """ Censors audio chunks in a continuous stream """
+
+        # Start thread that will analyze and censor recorded chunks
+        processing_thread = threading.Thread(target=self.run)
+        processing_thread.daemon = True
+        processing_thread.start()
+
+        try:
+            # Capture stream from Soundflower, play to speakers
+            with sd.Stream(device=(2, 1),
```
Comment (PatrickDuncan) [inline cleansio/censor/censor_realtime_mac.py]: The methods for looking for the desired sound device are already in the repo. Please fix the hard code

**Review by PatrickDuncan — CHANGES_REQUESTED** [review]

File: tests/censor/test_censor.py [hunk 15]
```diff
@@ -47,3 +48,5 @@ def test_censor():
     finally:
         # Cleanup
         os.remove(file_path_dup)
+
+test_censor()
```
Comment (PatrickDuncan) [inline tests/censor/test_censor.py]: Was this for debugging?

**Review by PatrickDuncan — CHANGES_REQUESTED** [review]
After that I'll merge

**Review by PatrickDuncan — APPROVED** [review]

**Pull request merged**